libvast/vast/qualified_record_field.hpp

```diff
@@ -24,23 +25,33 @@
 
 namespace vast {
 
-/// A standalone field of an event type, used to address an index column.
-/// Example: { "zeek.conn.id.orig_h", address_type{} }
+/// A standalone field of an event type, used to uniquely address an index
+/// column that may have the same field name across different event types.
+/// Example: { "zeek.conn", `id.orig_h", address_type{} }
 struct qualified_record_field
   : detail::totally_ordered<qualified_record_field> {
-  qualified_record_field(std::string full_name, type field_type)
-    : fqn{std::move(full_name)}, type{std::move(field_type)} {
-    // nop
+  // Required for serialization/deserialization.
+  qualified_record_field() = default;
+
+  /// Constructs a qualified record field by prepending the layout name to a
+  /// record field.
+  qualified_record_field(std::string record_name, record_field field)
+    : layout_name{std::move(record_name)},
+      field_name{std::move(field.name)},
+      type{std::move(field.type)} {
+    VAST_ASSERT(!layout_name.empty());
+    VAST_ASSERT(!field_name.empty());
   }
 
-  qualified_record_field(const std::string& layout_name,
-                         const record_field& field)
-    : fqn{layout_name + "." + field.name}, type{field.type} {
-    // nop
+  /// Retrieves the full-qualified name, i.e., the record typename concatenated
+  /// with the field name.
+  std::string fqn() const {
+    return layout_name + "." + field_name;
   }
 
-  std::string fqn; ///< The field name prepended with the record type name.
-  vast::type type; ///< The type of the field.
+  std::string layout_name; ///< The name of the layout.
+  std::string field_name;  ///< The name of the field.
```

Comment:
We don't use these name parts individually anywhere, so I don't understand the reasoning behind this change. I would also like to make sure this does not come with a negative impact on performance.

Reply:
What is the expensive operation you are concerned about? It doesn't change the hashing performance by much, so the only one I could think of is string concatenation when FQN is needed. My intuition is that this is not at all on the hot path, compared to whats going to happen with all the payload data.

> We don't use these name parts individually anywhere

That's not true, just look at `layout_name` in the diff. The separation of these fields is important so that we can use `layout_name` separately. Otherwise it's impossible to get back to the original `record_field`. We have a lot of code that dispatches by type name, which makes `qualified_record_field` much more flexible.

Reply:
I missed that use above. See my other comment.